src/plugins/visualizations/public/vis.ts

```diff
@@ -113,7 +113,16 @@ export class Vis<TVisParams = VisParams> {
     return defaults({}, cloneDeep(params ?? {}), cloneDeep(this.type.visConfig?.defaults ?? {}));
   }
 
-  async setState(state: PartialVisState) {
+  async setState(inState: PartialVisState) {
+    let state = inState;
+
+    const { updateVisTypeOnParamsChange } = this.type;
+    const newType = updateVisTypeOnParamsChange && updateVisTypeOnParamsChange(state.params);
+    if (newType) {
+      state = JSON.parse(JSON.stringify(inState));
+      state.params.type = state.type = newType;
```

Comment:
nit. Just prefer spread over JSON stringify/parse in this case
```suggestion
      state = {
        ... inState,
        type: newType,
        params: { ... inState.params, type: newType }
     };
```

Reply:
Nice.